### ISSUE TITLE
Discord adapter urgency detection: replace 3-word regex with layered signals (sigil + sender map + channel default)

### ISSUE THREAD

**Issue by jeffrichley** [jeffrichley/agent_core#38]
## Symptom (live, 2026-05-06)

Jeff sent Pepper a casual message about Lancaster trip planning:

> "ok, more on the ideas for the lancaster trip. right now we are looking at doing..."

The Discord adapter promoted the inbound `TextMessage` envelope to `urgency: "red"` because the regex matched the word "now" in the phrase "right **now** we are looking at...". Pepper read it as red-priority and surfaced "🔴 Red urgency. Reading immediately."

Trip planning is not red.

## Root cause

```python
# packages/agent-core-discord/src/agent_core_discord/access.py:34
urgency_red_regex: str = r"(?i)\b(urgent|now|stop)\b"
```

The shipped default uses a 3-word regex on raw message content. Two of the three words ("now", "stop") are among the most common English words. False positives are inevitable, the operator has to discover them by being annoyed, and tightening the regex per-deployment doesn't compose with anything else.

This pattern works for the demo case ("URGENT: server down!") and breaks for everything else.

## Why "tighten the regex" is the wrong fix

Regex on free-form natural-language content is structurally fragile:

- **Locale-specific** — English-only. Agents serving non-English users get no coverage at all.
- **Brittle to paraphrase** — "the system is hosed" should be red, "I might need that now" should be green. No regex can split these without a dictionary the size of WordNet.
- **No context** — a "stop" in a code comment shouldn't fire; a "stop" in an emergency message should.
- **Configuration-by-regex is operator-hostile** — every false positive forces the operator to learn the regex feature, write a tighter pattern, override the default, restart the daemon. Compounding cost.

We can do better than tuning regex.

### A. Explicit sigil prefix (deterministic, user-controlled)

Sender prefixes their message with a sigil to set urgency explicitly:

| Prefix | Urgency |
|---|---|
| `!` | red |
| `?` | yellow |
| (none) | green |
| `~` | green (explicit, suppresses any other rule) |

Pros: zero false positives, no NLP. Author-controlled, which is the right authority for urgency. Locale-independent.

Cons: requires user discipline. Muscle memory takes a few weeks. Doesn't help when urgency is *implicit* (the boss texting "where are we" at 11pm probably wants attention even without `!`).

### B. Sender-map defaults (relationship-aware)

Operator-configured map of user IDs → default urgency floor:

```json
{
  "senderUrgencyFloor": {
    "251892520399667200": "yellow",      // Jeff: yellow floor
    "<oncall-userid>":     "red",         // on-call: red floor
    "<friend-userid>":     "green"        // chitchat: green floor
  }
}
```

Combined with sigil prefix, the floor is a baseline and the sigil can promote (never demote). Pros: encodes "boss messages are pre-attention" without keyword brittleness.

Cons: relationship-management config grows over time. Doesn't catch the on-call user texting trip planning during the workday.

### C. Channel-of-origin defaults (context-aware)

Different channels have different urgency baselines:

```json
{
  "channelUrgencyDefault": {
    "<incidents-channel-id>": "red",
    "<ops-channel-id>":       "yellow",
    "<chitchat-channel-id>":  "green"
  }
}
```

Pros: incident-channel messages are red by default; chitchat is green. Matches actual operator mental model.

Cons: channel proliferation makes maintenance painful; doesn't help inside a multipurpose channel.

### D. Embedding-similarity to canonical samples (semantic, no regex)

Operator provides 5-10 example messages per urgency tier:

```json
{
  "urgencySamples": {
    "red": [
      "the server is down",
      "stop everything I need help",
      "everything's on fire"
    ],
    "yellow": [
      "where are we on this",
      "can you take a look at X"
    ]
  }
}
```

At inbound time, embed the message and find nearest-neighbor among samples. Use that bucket. Pros: handles paraphrases; no per-agent regex tuning; samples are easier to maintain than regex (operators THINK in examples, not patterns). Single embedding-model dependency. Locale-portable if the embedding model is multilingual.

Cons: cold-start needs samples; embedding model dependency (latency + memory). For a personal agent this is overkill, but the *architecture* is interesting because it generalizes.

### E. Tiny distilled classifier (small neural net)

Fine-tune a small model (~5M-50M params, e.g., DistilBERT-class or even a custom RNN) on a labeled corpus of (message, urgency) pairs. Run on every inbound. Pros: handles nuance, fast (<10ms inference), no per-message LLM cost.

Cons: training data has to come from somewhere; gets stale; per-deployment customization requires retraining. Most operators won't have a labeled corpus. Probably overkill for v1; would be a downstream evolution of (D).

### F. Self-classify by the receiving agent (LLM-as-classifier)

The adapter doesn't classify at all — every inbound is `urgency: "green"`. The agent itself reads the message and decides whether to escalate. The LLM the agent is already running classifies as a side-effect of reading the message.

Pros: leverages full LLM context (the agent knows what it's working on, who the sender is, time of day, recent activity). Most accurate. No new dependencies.

Cons: latency on inbound (one LLM round-trip before triaging); cost per inbound (fractions of a cent at small scale). Most importantly: changes the architectural contract — currently the adapter sets urgency before delivery, agents downstream trust it. Moving classification into the agent flips that.

This is also the closest to how a human triage assistant works — read the message, then decide if it's important.

### G. Stacked / layered (recommended for v2)

The sharpest design composes the cheap deterministic signals first, then falls back to richer methods only when needed:

```
1. Explicit sigil in message?     → use it (deterministic override)
2. Sender in red allowlist?       → red
3. Channel default?               → use as floor
4. (optional) Embedding similarity to samples → bump if matches red samples
5. Otherwise                      → green
```

Pros: cheap-first, escalate-when-needed; each layer is independently configurable; no single layer has to be perfect. Author keeps control via sigil; operator keeps control via per-sender / per-channel maps; semantic catch-all via samples.

Cons: more configuration surface to learn — but each layer is independently optional.

## Recommendation

**Land (A) + (B) + (C) as v2.** Drop the regex entirely from the default config. Three layered, deterministic signals (prefix, sender, channel) give 90% of the value with 10% of the complexity. Operators can opt into (D) embedding samples later if they want semantic robustness.

Defer (E) and (F) — interesting directions but premature.

## Tests to add

- `test_sigil_prefix_promotes_urgency` — `!hello` → red, `?status` → yellow, plain → green
- `test_sigil_prefix_strips_from_message_content` — `!server down` becomes payload "server down" with urgency=red, NOT "!server down" with red
- `test_sender_urgency_floor_applied` — message from configured user gets the floor as baseline
- `test_channel_urgency_default_applied` — message in configured channel gets the default
- `test_layered_urgency_max_of_signals` — sigil (`?` yellow) + sender (red floor) → red wins
- `test_no_signals_falls_through_to_green` — plain message with no overrides is green

## Out of scope

- Multi-locale urgency keyword expansion (replaced by approach D's samples or by removing the regex entirely).
- Notification / wake-builder behavior changes (issue #33 separately).
- Cross-platform parity (this issue is Discord-specific; Slack / email adapters can adopt the same approach when they implement urgency).

## Migration

Existing deployments using `urgency_red_regex` continue to work (the field stays on `AccessConfig`, just defaults to empty string instead of the 3-word pattern). Operators who want regex behavior can opt in. New deployments get sigil + sender + channel out of the box.

## Related

- 2026-05-06 false positive: Jeff's Lancaster trip message → red because of "now". Surfaced during Pepper post-cutover use.
- Issue #33 — wake-builder uses urgency to compute `urgency_max`; better classification → better wake signal.
- Issue #34 — different access-control gap (scheduler), same family of v1-bus permissive defaults.
- Issue #36 — Discord deny_channels (different concern, also surfaced during 2026-05-06).
